extensions/sql-database-projects/src/models/project.ts

```diff
@@ -307,22 +307,31 @@ export class Project implements ISqlProject {
 	/**
 	 * Adds a folder to the project, and saves the project file
 	 * @param relativeFolderPath Relative path of the folder
+	 * @param ignoreDuplicate
+	 *	Flag that indicates whether duplicate entries should be ignored. If flag is set to `true` and item already
+	 *	exists in the project file, then existing entry will be returned.
 	 */
-	public async addFolderItem(relativeFolderPath: string): Promise<FileProjectEntry> {
+	public async addFolderItem(relativeFolderPath: string, ignoreDuplicate?: boolean): Promise<FileProjectEntry> {
 		const absoluteFolderPath = path.join(this.projectFolderPath, relativeFolderPath);
+		const normalizedRelativeFolderPath = utils.convertSlashesForSqlProj(relativeFolderPath);
 
 		// check if folder already has been added to sqlproj
-		if (this.files.find(f => f.relativePath.toLowerCase() === relativeFolderPath.toLowerCase())) {
-			throw new Error(constants.folderAlreadyAddedToProject((relativeFolderPath)));
+		const existingEntry = this._files.find(f => f.relativePath.toUpperCase() === normalizedRelativeFolderPath.toUpperCase());
```

Comment:
I changed comparison to use upper-case to make sure non-Latin characters are compared properly.

Reply:
Conversion should also be done outside of the comparison delegate, as it is a constant value, but the code is already plagued with bad coding practices, so I went for a least destructive change to simplify diff.

Reply:
I also don't know if there was a reason to use property getter (`this.files`), instead of a field directly (`this._files`). I changed it to use the field, as done in other places in this file.

Reply:
I'm not really sure how .NET is in play here, as this is a culture-related issue, where in some cultures there might be two upper-case characters (accented and not), but only one low-case character, thus any implementation of to-lower will lose fidelity. I don't think `string.normalize()` should be used here either. If we want a CI AS comparison, then `Intl.Collator` with appropriate configuration should be used.

In reality though, what we want is not a CI AS comparison, but to determine whether this file on disk is already added to the project and this is a little bit more complicated than just running a string comparison. First of all, it is based on the file system where the file is stored - some allow same name with different case and some do not. Then there is an extra normalization necessary for folders to make sure correct path separator is used and trailing separators are present/trimmed for folders. Ultimately you probably want to find a file/folder on disk using `fs.stat` or something and extract canonical file name from it for comparison. This further hints the necessity of this comparison logic to be present here, rather than duplicated within each and every consumer.

Reply:
> Depends on the codebase's priorities. For performance, yes. For concise, more-readable code, something like this is better.

:) This is probably the most bizarre statement I've ever seen. I'd pay to see a conversation with a customer where we explain to them that our tools are running slow, because instead our code looks "pretty" and that what matters the most.
Not trying to diminish the value of readable and self-explanatory code, but I don't think there is ever a choice for "codebase's priorities" between user experience (performance, in this case) and code readability. Usually you would decide on the target user experience and try to achieve it with the most "concise, more-readable code" possible, but not the opposite. The point can be made that the performance is "good enough", but I think this is one of the cases where improving it doesn't require virtually any effort. Otherwise (i.e. "we must have it written this way, otherwise I don't understand what is going on"), it probably deserve a change in `tsc` to improve the DFA and optimize this during transpilation.